pkg/host/configvalidation.go

```diff
@@ -0,0 +1,312 @@
+
+	"github.com/Mellanox/nic-configuration-operator/api/v1alpha1"
+	"github.com/Mellanox/nic-configuration-operator/pkg/consts"
+	"github.com/Mellanox/nic-configuration-operator/pkg/types"
+)
+
+type configValidation interface {
+	// ConstructNvParamMapFromTemplate translates a configuration template into a set of nvconfig parameters
+	// operates under the assumption that spec validation was already carried out
+	ConstructNvParamMapFromTemplate(
+		device *v1alpha1.NicDevice, defaultValues map[string]string) (map[string]string, error)
+	// ValidateResetToDefault checks if device's nv config has been reset to default in current and next boots
+	// returns bool - need to perform reset
+	// returns bool - reboot required
+	// returns error - if an error occurred during validation
+	ValidateResetToDefault(nvConfig types.NvConfigQuery) (bool, bool, error)
+	// AdvancedPCISettingsEnabled returns true if ADVANCED_PCI_SETTINGS param is enabled for current config
+	AdvancedPCISettingsEnabled(currentConfig map[string]string) bool
+	// RuntimeConfigApplied checks if desired runtime config is applied
+	RuntimeConfigApplied(device *v1alpha1.NicDevice) (bool, error)
+	// CalculateDesiredRuntimeConfig returns desired values for runtime config
+	// returns int - maxReadRequestSize
+	// returns string - qos trust mode
+	// returns string - qos pfc settings
+	// returns err - an error occurred
+	CalculateDesiredRuntimeConfig(device *v1alpha1.NicDevice) (int, string, string)
+}
+
+type configValidationImpl struct {
+	utils HostUtils
+}
+
+func nvParamLinkTypeFromName(linkType string) string {
+	if linkType == consts.Infiniband {
+		return consts.NvParamLinkTypeInfiniband
+	} else if linkType == consts.Ethernet {
+		return consts.NvParamLinkTypeEthernet
+	}
+
+	return ""
+}
+
+func applyDefaultNvConfigValueIfExists(
+	paramName string, desiredParameters map[string]string, defaultValues map[string]string) {
+	defaultValue, found := defaultValues[paramName]
+	// Default values might not yet be available if ENABLE_PCI_OPTIMIZATIONS is disabled
+	if found {
+		desiredParameters[paramName] = defaultValue
+	}
+}
+
+// ConstructNvParamMapFromTemplate translates a configuration template into a set of nvconfig parameters
+// operates under the assumption that spec validation was already carried out
+func (v *configValidationImpl) ConstructNvParamMapFromTemplate(
+	device *v1alpha1.NicDevice, defaultValues map[string]string) (map[string]string, error) {
+	desiredParameters := map[string]string{}
+
+	template := device.Spec.Configuration.Template
+	secondPortPresent := len(device.Status.Ports) > 1
+	pciAddr := device.Status.Ports[0].PCI
+
+	if template.NumVfs == 0 {
+		desiredParameters[consts.SriovEnabledParam] = consts.NvParamFalse
+		desiredParameters[consts.SriovNumOfVfsParam] = "0"
+	} else {
+		desiredParameters[consts.SriovEnabledParam] = consts.NvParamTrue
+		desiredParameters[consts.SriovNumOfVfsParam] = strconv.Itoa(template.NumVfs)
+	}
```

Comment:
nit: to follow the go-style

```suggestion
	desiredParameters[consts.SriovEnabledParam] = consts.NvParamFalse
	desiredParameters[consts.SriovNumOfVfsParam] = "0"
	if template.NumVfs > 0 {
		desiredParameters[consts.SriovEnabledParam] = consts.NvParamTrue
		desiredParameters[consts.SriovNumOfVfsParam] = strconv.Itoa(template.NumVfs)
	}
```